src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerGroupedByLocus.java

```diff
@@ -0,0 +1,203 @@
+                        apply(variant,
+                                Collections.singletonList(variant),
+                                new ReadsContext(reads, variantInterval, readFilter),
+                                new ReferenceContext(reference, variantInterval),
+                                new FeatureContext(features, variantInterval));
+
+                        progressMeter.update(variantInterval);
+                    });
+        }
+    }
+
+    protected Stream<VariantContext> getTransformedVariantStream(final Spliterator<VariantContext> source, final VariantFilter filter) {
+        final VariantTransformer preTransformer  = makePreVariantFilterTransformer();
+        final VariantTransformer postTransformer = makePostVariantFilterTransformer();
+        return StreamSupport.stream(source, false)
+                .map(preTransformer)
+                .filter(filter)
+                .map(postTransformer);
+    }
+
+    // Return a Stream of SimpleInterval covering the entire territory sketched out by requestedIntervals
+    private Stream<SimpleInterval> getLocusStream(final List<SimpleInterval> requestedIntervals) {
+        final Iterable<SimpleInterval> iterable = () -> new IntervalLocusIterator(requestedIntervals.iterator());
+        return StreamSupport.stream(iterable.spliterator(), false);
+    }
+
+    /**
+     * Process an individual variant. Must be implemented by tool authors.
+     * In general, tool authors should simply stream their output from apply(), and maintain as little internal state
+     * as possible.
+     *
+     * @param loc
+     * @param variants Current variants being processed.
+     * @param readsContext Reads overlapping the current variant. Will be an empty, but non-null, context object
+     *                     if there is no backing source of reads data (in which case all queries on it will return
+     *                     an empty array/iterator)
+     * @param referenceContext Reference bases spanning the current variant. Will be an empty, but non-null, context object
+     *                         if there is no backing source of reference data (in which case all queries on it will return
+     *                         an empty array/iterator). Can request extra bases of context around the current variant's interval
+     *                         by invoking {@link ReferenceContext#setWindow}
+     *                         on this object before calling {@link ReferenceContext#getBases}
+     * @param featureContext Features spanning the current variant. Will be an empty, but non-null, context object
+     *                       if there is no backing source of Feature data (in which case all queries on it will return an
+     *                       empty List).
+     */
+    public abstract void apply(Locatable loc, List<VariantContext> variants, ReadsContext readsContext, ReferenceContext referenceContext, FeatureContext featureContext );
```

Comment:
This doc needs a big callout about the mode switching and what happens in each mode.

Reply:
Done.